clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -4676,12 +4676,21 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
     if (HasFPGA)
       CmdArgs.push_back("-fsycl-disable-range-rounding");
 
+    const auto &DeviceTriple = getToolChain().getTriple();
+    const auto &DeviceSubArch = DeviceTriple.getSubArch();
     // Enable generation of USM address spaces for FPGA.
     // __ENABLE_USM_ADDR_SPACE__ will be used during compilation of SYCL headers
-    if (getToolChain().getTriple().getSubArch() ==
-        llvm::Triple::SPIRSubArch_fpga)
+    if (DeviceSubArch == llvm::Triple::SPIRSubArch_fpga)
       CmdArgs.push_back("-D__ENABLE_USM_ADDR_SPACE__");
 
+    // Use native FP atomics on supported targets by default.
+    // TODO: Currently, only OpenCL CPU and Intel GPU support the functions.
+    // Allow other targets through check once/if support for
+    // SPV_EXT_shader_atomic_float* extensions lands there.
+    if (DeviceTriple.isSPIR() &&
+        DeviceSubArch != llvm::Triple::SPIRSubArch_fpga)
+      CmdArgs.push_back("-DSYCL_USE_NATIVE_FP_ATOMICS");
```

Comment:
Typically macro settings are done in the FE, why not do it there instead of the driver?

Reply:
That's a good point; I'll try moving the logic there. I've initially followed the example of `__ENABLE_USM_ADD_SPACE___`, but as of now, it looks like its initialization might also be worth moving into `InitPreprocessor.cpp` (I'll make a follow-up PR).

Reply:
Done in https://github.com/intel/llvm/pull/3869/commits/a4685ba55be39e6beeac390b7467203a71513a08. I've also created https://github.com/intel/llvm/issues/3972 on myself to address other inconsistencies with device-side macro initializations.